### PR TITLE
Adding Azure.Identity.BrokeredAuthentication package (internals visible)

### PR DESCRIPTION
**Pull request by schaabs**

### CODE REVIEW

File: sdk/identity/Azure.Identity.BrokeredAuthentication/README.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
+# Azure Identity Brokered Authentication client library for .NET
+ The library extends the Azure.Identity library to provide authentication broker support. It includes the necessary dependancies, and provides the `InteractiveBrowserCredentialBrokerOptions` class. This optoins class can be used to create an `InteractiveBrowserCredential` capable of using the sytem authentication broker in lieue of the system browser when available.
```
Comment (christothes) [inline sdk/identity/Azure.Identity.BrokeredAuthentication/README.md:2]:
```suggestion
 The library extends the Azure.Identity library to provide authentication broker support. It includes the necessary dependencies and provides the `InteractiveBrowserCredentialBrokerOptions` class. This options class can be used to create an `InteractiveBrowserCredential` capable of using the system authentication broker in lieu of the system browser when available.  
```

File: sdk/identity/Azure.Identity.BrokeredAuthentication/README.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  [Source code][source] | [Package (nuget)][package] | [API reference documentation][identity_api_docs] | [Azure Active Directory documentation][aad_doc]
+
+## Getting started
+
+### Install the package
+
+Install the Azure Identity client library for .NET with [NuGet][nuget]:
+
+```PowerShell
+Install-Package Azure.Identity.BrokeredAuthentication
```
Comment (christothes) [inline sdk/identity/Azure.Identity.BrokeredAuthentication/README.md:13]: do we need `--pre-release`?

File: sdk/identity/Azure.Identity.BrokeredAuthentication/tests/Azure.Identity.BrokeredAuthentication.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,24 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(RequiredTargetFrameworks);netcoreapp3.1;net461</TargetFrameworks>
```
Comment (christothes) [inline sdk/identity/Azure.Identity.BrokeredAuthentication/tests/Azure.Identity.BrokeredAuthentication.Tests.csproj:4]: are the extra targets necessary? I thought we'd get them all from RequiredTargetFrameworks

File: sdk/identity/Azure.Identity.BrokeredAuthentication/Directory.Build.props [hunk 4]
```diff
@@ -0,0 +1,6 @@
+﻿<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (christothes) [inline sdk/identity/Azure.Identity.BrokeredAuthentication/Directory.Build.props:1]: Is this file necessary?

File: sdk/identity/Azure.Identity.BrokeredAuthentication/src/Azure.Identity.BrokeredAuthentication.csproj [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    <!-- <ApiCompatVersion>1.0.0</ApiCompatVersion> -->
+    <PackageTags>Microsoft Azure Identity Broker;$(PackageCommonTags)</PackageTags>
+    <TargetFrameworks>$(RequiredTargetFrameworks);net461</TargetFrameworks>
+    <NoWarn>$(NoWarn);3021</NoWarn>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Azure.Core" />
+    <!--<PackageReference Include="Azure.Identity" />-->
+    <!-- This should be changed back to a package reference once the required changes in Azure.Identity have shipped in the next GA -->
```
Comment (christothes) [inline sdk/identity/Azure.Identity.BrokeredAuthentication/src/Azure.Identity.BrokeredAuthentication.csproj:16]: Should we create an issue to track this?

File: sdk/identity/Azure.Identity.BrokeredAuthentication/src/Azure.Identity.BrokeredAuthentication.csproj [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    <ProjectReference Include="..\..\Azure.Identity\src\Azure.Identity.csproj" />
+    <PackageReference Include="System.Memory" />
+    <PackageReference Include="System.Text.Json" />
+    <PackageReference Include="System.Threading.Tasks.Extensions" />
+    <PackageReference Include="Microsoft.Identity.Client" />
+    <PackageReference Include="Microsoft.Identity.Client.Extensions.Msal" />
+    <PackageReference Include="System.Security.Cryptography.ProtectedData" />
+  </ItemGroup>
+  <!--Only Add Microsoft.Identity.Client.Desktop when compiled for desktop or net core 3.1-->
```
Comment (christothes) [inline sdk/identity/Azure.Identity.BrokeredAuthentication/src/Azure.Identity.BrokeredAuthentication.csproj:25]: This comment doesn't seem to match the condition.

File: sdk/identity/Azure.Identity.BrokeredAuthentication/src/InteractiveBrowserCredentialBrokerOptions.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+using Microsoft.Identity.Client;
+#if (NETFRAMEWORK)
+using Microsoft.Identity.Client.Desktop;
+#endif
+
+namespace Azure.Identity.BrokeredAuthentication
+{
+    /// <summary>
+    /// Options to configure the <see cref="InteractiveBrowserCredential"/> to use the system authentication broker in lieue of the system browser if available.
```
Comment (christothes) [inline sdk/identity/Azure.Identity.BrokeredAuthentication/src/InteractiveBrowserCredentialBrokerOptions.cs:12]:
```suggestion
    /// Options to configure the <see cref="InteractiveBrowserCredential"/> to use the system authentication broker in lieu of the system browser if available.
```

File: sdk/identity/Azure.Identity.sln [hunk 8]
```diff
@@ -3,23 +3,33 @@ Microsoft Visual Studio Solution File, Format Version 12.00
...
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.Identity.Tests", "Azure.Identity\tests\Azure.Identity.Tests.csproj", "{CD0BE1AA-F99E-48E7-9FFD-F1B288E62B4F}"
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution Items", "{8F748B65-7779-4A71-B2EC-9BA8B9526AB0}"
 	ProjectSection(SolutionItems) = preProject
-		CHANGELOG.md = CHANGELOG.md
-		..\ci.yml = ..\ci.yml
-		README.md = README.md
-		..\tests.yml = ..\tests.yml
+		ci.yml = ci.yml
+		tests.yml = tests.yml
```
Comment (christothes) [inline sdk/identity/Azure.Identity.sln:13]: Do we need this and the other SolutionItems stuff?

File: sdk/identity/Azure.Identity/src/Azure.Identity.csproj [hunk 9]
```diff
@@ -7,8 +7,9 @@
     <ApiCompatVersion>1.5.0</ApiCompatVersion>
     <PackageTags>Microsoft Azure Identity;$(PackageCommonTags)</PackageTags>
     <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
-    <NoWarn>$(NoWarn);3021</NoWarn>
+    <NoWarn>$(NoWarn);3021;AZC0011</NoWarn>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <DefineConstants>IDENTITYCORE</DefineConstants>
```
Comment (christothes) [inline sdk/identity/Azure.Identity/src/Azure.Identity.csproj:12]: Do we still need this?

File: sdk/identity/Azure.Identity/src/MsalPublicClient.cs [hunk 10]
```diff
@@ -53,6 +56,11 @@ protected virtual ValueTask<IPublicClientApplication> CreateClientCoreAsync(stri
                 pubAppBuilder.WithClientCapabilities(clientCapabilities);
             }
 
+            if (_beforeBuildClient != null)
```
Comment (christothes) [inline sdk/identity/Azure.Identity/src/MsalPublicClient.cs:59]: Should we add a test for this?

File: sdk/identity/Azure.Identity/src/Properties/AssemblyInfo.cs [hunk 11]
```diff
@@ -5,5 +5,6 @@
 
 [assembly: InternalsVisibleTo("Microsoft.Extensions.Azure.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100d15ddcb29688295338af4b7686603fe614abd555e09efba8fb88ee09e1f7b1ccaeed2e8f823fa9eef3fdd60217fc012ea67d2479751a0b8c087a4185541b851bd8b16f8d91b840e51b1cb0ba6fe647997e57429265e85ef62d565db50a69ae1647d54d7bd855e4db3d8a91510e5bcbd0edfbbecaa20a7bd9ae74593daa7b11b4")]
 [assembly: InternalsVisibleTo("Azure.Identity.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100d15ddcb29688295338af4b7686603fe614abd555e09efba8fb88ee09e1f7b1ccaeed2e8f823fa9eef3fdd60217fc012ea67d2479751a0b8c087a4185541b851bd8b16f8d91b840e51b1cb0ba6fe647997e57429265e85ef62d565db50a69ae1647d54d7bd855e4db3d8a91510e5bcbd0edfbbecaa20a7bd9ae74593daa7b11b4")]
+[assembly: InternalsVisibleTo("Azure.Identity.BrokeredAuthentication, PublicKey=0024000004800000940000000602000000240000525341310004000001000100097ad52abbeaa2e1a1982747cc0106534f65cfea6707eaed696a3a63daea80de2512746801a7e47f88e7781e71af960d89ba2e25561f70b0e2dbc93319e0af1961a719ccf5a4d28709b2b57a5d29b7c09dc8d269a490ebe2651c4b6e6738c27c5fb2c02469fe9757f0a3479ac310d6588a50a28d7dd431b907fd325e18b9e8ed")]
```
Comment (weshaggard) [inline sdk/identity/Azure.Identity/src/Properties/AssemblyInfo.cs:8]: Why do we need to expose these internal apis? How do we plan on ensuring we don't break this internal contract? Versioning libraries can become very difficult when there is this internal contract.

File: sdk/identity/Azure.Identity.BrokeredAuthentication/tests/Azure.Identity.BrokeredAuthentication.Tests.csproj (hunk 3 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity.BrokeredAuthentication/tests/Azure.Identity.BrokeredAuthentication.Tests.csproj:4]: I don't think they are. By default, I believe only netstandard2.0 is targeted. We need to build for these other targets as well to since we have #if directives which are conditionally compiled in for these frameworks.

File: sdk/identity/Azure.Identity.BrokeredAuthentication/Directory.Build.props (hunk 4 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity.BrokeredAuthentication/Directory.Build.props:1]: I believe it serves to pull shared properties from parent folders, but I'm not sure. I was following the folder structure of our existing project.

File: sdk/identity/Azure.Identity.sln (hunk 8 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity.sln:13]: I had them in there for convenience, but I can remove them.

File: sdk/identity/Azure.Identity.BrokeredAuthentication/tests/Azure.Identity.BrokeredAuthentication.Tests.csproj (hunk 3 above)
Comment (christothes) [reply sdk/identity/Azure.Identity.BrokeredAuthentication/tests/Azure.Identity.BrokeredAuthentication.Tests.csproj]: But if you remove them, we don't still build those framework versions?

File: sdk/identity/Azure.Identity.BrokeredAuthentication/Directory.Build.props (hunk 4 above)
Comment (christothes) [reply sdk/identity/Azure.Identity.BrokeredAuthentication/Directory.Build.props:1]: I have seen it missing in newer projects, so I think if it just does the default, it can be removed.

Comment (schaabs) [reply sdk/identity/Azure.Identity.BrokeredAuthentication/Directory.Build.props:1]: I noticed that including the Directory.Build.props modified my project. It pulled the README.md and CHANGELIST.md into the project without me having to explicitly add item.

File: sdk/identity/Azure.Identity/src/MsalPublicClient.cs (hunk 10 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/MsalPublicClient.cs:59]: added.

File: sdk/identity/Azure.Identity.BrokeredAuthentication/tests/Azure.Identity.BrokeredAuthentication.Tests.csproj (hunk 3 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity.BrokeredAuthentication/tests/Azure.Identity.BrokeredAuthentication.Tests.csproj]: No if we don't explicitly include them the package will only have a single assembly for netstandard2.0.

File: sdk/identity/Azure.Identity.BrokeredAuthentication/src/Azure.Identity.BrokeredAuthentication.csproj (hunk 5 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity.BrokeredAuthentication/src/Azure.Identity.BrokeredAuthentication.csproj:16]: https://github.com/Azure/azure-sdk-for-net/issues/26879

File: sdk/identity/Azure.Identity.BrokeredAuthentication/README.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+  [Source code][source] | [Package (nuget)][package] | [API reference documentation][identity_api_docs] | [Azure Active Directory documentation][aad_doc]
+
+## Getting started
+
+### Install the package
+
+Install the Azure Identity client library for .NET with [NuGet][nuget]:
+
+```PowerShell
+Install-Package Azure.Identity.BrokeredAuthentication -IncludePrerelease
```
Comment (christothes) [inline sdk/identity/Azure.Identity.BrokeredAuthentication/README.md:13]:
```suggestion
dotnet add package Azure.Identity.BrokeredAuthentication --prerelease
```

**Review by christothes — APPROVED** [review]
